sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-retryablehttp"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/hashicorp/vault/sdk/helper/certutil"
+	"golang.org/x/crypto/ocsp"
+)
+
+// FailOpenMode is OCSP fail open mode. FailOpenTrue by default and may
+// set to ocspModeFailClosed for fail closed mode
+type FailOpenMode uint32
+
+type requestFunc func(method, urlStr string, body interface{}) (*retryablehttp.Request, error)
+
+type clientInterface interface {
+	Do(req *retryablehttp.Request) (*http.Response, error)
+}
+
+const (
+	httpHeaderContentType   = "Content-Type"
+	httpHeaderAccept        = "accept"
+	httpHeaderContentLength = "Content-Length"
+	httpHeaderHost          = "Host"
+	ocspRequestContentType  = "application/ocsp-request"
+	ocspResponseContentType = "application/ocsp-response"
+)
+
+const (
+	ocspFailOpenNotSet FailOpenMode = iota
+	// FailOpenTrue represents OCSP fail open mode.
+	FailOpenTrue
+	// FailOpenFalse represents OCSP fail closed mode.
+	FailOpenFalse
+)
+
+const (
+	ocspModeFailOpen   = "FAIL_OPEN"
+	ocspModeFailClosed = "FAIL_CLOSED"
+	ocspModeInsecure   = "INSECURE"
+)
+
+const ocspCacheKey = "ocsp_cache"
+
+const (
+	// defaultOCSPResponderTimeout is the total timeout for OCSP responder.
+	defaultOCSPResponderTimeout = 10 * time.Second
+)
+
+const (
+	// cacheExpire specifies cache data expiration time in seconds.
+	cacheExpire = float64(24 * 60 * 60)
+)
+
+type ocspCachedResponse struct {
+	time       float64
+	producedAt float64
+	thisUpdate float64
+	nextUpdate float64
+	status     ocspStatusCode
+}
+
+type Client struct {
+	// caRoot includes the CA certificates.
+	caRoot map[string]*x509.Certificate
+	// certPool includes the CA certificates.
+	certPool              *x509.CertPool
+	ocspResponseCache     *lru.TwoQueueCache
+	ocspResponseCacheLock sync.RWMutex
+	// cacheUpdated is true if the memory cache is updated
+	cacheUpdated bool
+	logFactory   func() hclog.Logger
+}
+
+type ocspStatusCode int
+
+type ocspStatus struct {
+	code ocspStatusCode
+	err  error
+}
+
+const (
+	ocspSuccess                ocspStatusCode = 0
+	ocspStatusGood             ocspStatusCode = -1
+	ocspStatusRevoked          ocspStatusCode = -2
+	ocspStatusUnknown          ocspStatusCode = -3
+	ocspStatusOthers           ocspStatusCode = -4
+	ocspFailedDecomposeRequest ocspStatusCode = -5
+	ocspInvalidValidity        ocspStatusCode = -6
+	ocspMissedCache            ocspStatusCode = -7
+	ocspCacheExpired           ocspStatusCode = -8
+)
+
+// copied from crypto/ocsp.go
+type certID struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	NameHash      []byte
+	IssuerKeyHash []byte
+	SerialNumber  *big.Int
+}
+
+// cache key
+type certIDKey struct {
+	HashAlgorithm crypto.Hash
```

Comment:
Per discussion on Slack, we can probably remove this field and clarify that `NameHash` is the Issuer's Name (for uniqueness) and the `SerialNumber` is the leaf's serial number; combination of `(IssuerName, IssuerKey)` should uniquely identify the issuer, and `LeafSerialNumber` should uniquely identify the leaf within that parent issuer.